gh-115999: Specialize `STORE_ATTR` in free-threaded builds.

nascheme opened the pull request:
* Fix thread safety issues with specialized opcodes (`STORE_ATTR_INSTANCE_VALUE`, `STORE_ATTR_SLOT`, `STORE_ATTR_WITH_HINT`).  Need a combination of locks and atomics to be safe.
* Fix thread safety issues with `_Py_Specialize_StoreAttr` .  Avoid using borrowed references.  Save and store the `tp_version_tag` from the beginning of the specialization process since it might change.  Use helper functions to update opcode.
* Add unit tests to ensure specialization is happening in free-threaded builds.  Add addtional tests to try to trigger data races.


<!-- gh-issue-number: gh-115999 -->
* Issue: gh-115999
<!-- /gh-issue-number -->

nascheme committed:
Enable specialization of `STORE_ATTR` free-threaded builds.

colesbury commented on Python/bytecodes.c:
```diff
@@ -2359,16 +2361,33 @@ dummy_func(
+            if (dict == NULL) {
+                UNLOCK_OBJECT(dict);
+                DEOPT_IF(true);
+            }
```
There's already a `dict == NULL` check above. I think this should be:

```
DEOPT_IF(!LOCK_OBJECT(dict));
```

nascheme replied:
Fixed

colesbury commented on Python/bytecodes.c:
```diff
@@ -2326,25 +2326,27 @@ dummy_func(
             PyObject *owner_o = PyStackRef_AsPyObjectBorrow(owner);
-
+            DEOPT_IF(!LOCK_OBJECT(owner_o));
+            if (_PyObject_GetManagedDict(owner_o) == NULL) {
```
The condition here doesn't look right. Previously we had:

```
assert(_PyObject_GetManagedDict(owner_o) == NULL)
```

But now the code deopts if that's the case:

```
if (_PyObject_GetManagedDict(owner_o) == NULL) {
    UNLOCK_OBJECT(owner_o);
    DEOPT_IF(true);
}
```

I think we need to lock the object earlier, in `_GUARD_DORV_NO_DICT` or `_GUARD_TYPE_VERSION`. It's a bit awkward to hold the lock across uops, but I don't know of a better way.

I think we want to check the type version tag under the lock, so maybe `_GUARD_TYPE_VERSION` there should also be a `_GUARD_TYPE_VERSION_AND_LOCK`.

nascheme replied:
Fixed, I think.  I didn't use `_GUARD_TYPE_VERSION_AND_LOCK` for the `STORE_ATTR_WITH_HINT` op, only for `STORE_ATTR_INSTANCE_VALUE`.  You want to lock the dict in that case.  I think the way it's coded now there is a kind of race.  If another thread creates a descriptor between `_GUARD_TYPE_VERSION` and the `STORE_ATTR_WITH_HINT`, the descriptor will get ignored.  I think that's okay.  I think otherwise you would have to lock both the object and object dict before checking `tp_version_tag`.  The code in `dictobject.c` doesn't appear to do that.

mpage commented on Lib/test/test_opcache.py:
```diff
@@ -1375,6 +1375,74 @@ def send_yield_from():
+
+        set_slot()
+
+        #dis.dis(set_slot, adaptive=True)
```
Left over from debugging?

```suggestion
```

nascheme replied:
Removed.

mpage commented on Python/specialize.c:
```diff
@@ -932,6 +935,61 @@ analyze_descriptor(PyTypeObject *type, PyObject *name, PyObject **descr, int sto
+        return 0;
+    }
+    cache->index = (uint16_t)offset;
+    write_u32(cache->version, type->tp_version_tag);
```
I don't think it's safe to re-read and store the type version here. It's possible that the type was modified between the call to `analyze_descriptor` and here and a value was inserted into the MRO that would invalidate our analysis (e.g. a data descriptor). If we store the current type version the guard would pass and we wouldn't execute the data descriptor.

I think we need to write a version tag that we know to be consistent with our analysis. One way to do that would be to:

1. Add a `tp_version` out param to `analyze_descriptor`, use `PyType_LookupRefAndVersion` in `analyze_descriptor`, and set the out param using the result. 
2. Thread the version from (1) through `specialize_dict_access` down to both `specialize_dict_access_inline` and `specialize_dict_access_hint`.

nascheme replied:
Fixed by passing `tp_version` to `specialize_dict_access` and down.

mpage commented on Python/specialize.c:
```diff
@@ -932,6 +935,61 @@ analyze_descriptor(PyTypeObject *type, PyObject *name, PyObject **descr, int sto
+        return 0;
+    }
+    cache->index = (uint16_t)index;
+    write_u32(cache->version, type->tp_version_tag);
```
Same comment as above.

mpage commented on Python/specialize.c:
```diff
@@ -946,55 +1004,28 @@ specialize_dict_access(
     }
-    _PyAttrCache *cache = (_PyAttrCache *)(instr + 1);
     if (type->tp_flags & Py_TPFLAGS_INLINE_VALUES &&
         _PyObject_InlineValues(owner)->valid &&
```
I think we need to hold the critical section on `owner` across the `_PyObject_InlineValues(owner)->valid` check and the call to `specialize_dict_access_inline`.  Otherwise, I think it's possible that we race with someone invalidating the inline values between our check and when we take the critical section.

nascheme replied:
I think I've addressed this now following similar logic to `_PyObject_StoreInstanceAttribute`.  Lock the object and then get the managed dict again.  If it was created, abort the specialization (we lost race during locking).

Do we need to an additional check to ensure the inline values are still valid?  I'm not sure.

mpage commented on Python/specialize.c:
```diff
@@ -1258,11 +1291,12 @@ _Py_Specialize_StoreAttr(_PyStackRef owner_st, _Py_CODEUNIT *instr, PyObject *na
-    DescriptorClassification kind = analyze_descriptor(type, name, &descr, 1);
-    if (type_get_version(type, STORE_ATTR) == 0) {
+    uint32_t tp_version = type_get_version(type, STORE_ATTR);
+    if (tp_version == 0) {
```
I think it's possible that we end up failing to specialize more frequently by hoisting this. Previously, the call to `type_get_version` happened after we called `PyType_Lookup` in `analyze_descriptor`. One side effect of `PyType_Lookup` is that it will assign a version tag. I think it's pretty unlikely, but it's possible that we encounter a type without a version in this arrangement and fail to specialize, whereas we would succeed in the previous version because one of the calls to `PyType_Lookup` in `analyze_descriptor` would assign a type version.

The suggestion from above to use `PyType_LooupRefAndVersion` in `analyze_descriptor` is one possible solution.

mpage reviewed (commented):
Thanks for taking this on! This looks like a [regression](https://github.com/facebookexperimental/free-threading-benchmarking/tree/main/results/bm-20241211-3.14.0a2+-7e3de44) on the default build. I haven't had a chance to dig into it, but I suspect it might either be due to the check that Sam flagged in `_STORE_ATTR_INSTANCE_VALUE` or the change to when we read the type version in `_Py_Specialize_StoreAttr`. It looks like the richards benchmark is the most heavily affected, so that might be a good isolated benchmark to use for debugging.

nascheme committed:
Fix issues found in code review.

* Fix locking for `STORE_ATTR_INSTANCE_VALUE`.  Create
  `_GUARD_TYPE_VERSION_AND_LOCK` so that object stays locked and
  `tp_version_tag` cannot change.  Fix inverted logic bug that caused
  erroneous deopt.

* Fix locking for `_STORE_ATTR_WITH_HINT`.  Double check that
  `_PyObject_GetManagedDict()` hasn't disappeared since we locked the
  dict.

- Pass `tp_version_tag` to `specialize_dict_access()`, ensuring
  the version we store on the cache is the correct one (in case of
  it changing during the specalize analysis).

- Split `analyze_descriptor` into `analyze_descriptor_load` and
  `analyze_descriptor_store` since those don't share much logic.
  Add `descriptor_is_class` helper function.

- In `specialize_dict_access`, double check `_PyObject_GetManagedDict()`
  in case we race and dict was materialized before the lock.

nascheme committed:
Small optimization for STORE_ATTR specialize.

If the type is new and a version tag hasn't yet been assigned, we would
fail to specialize it.  Use `_PyType_LookupRefAndVersion()` instead of
`type_get_version()`, which will assign a version.

nascheme committed:
Fix race in specialize_dict_access_inline().

Use provided value of `tp_version` to store in cache.

colesbury commented on Python/bytecodes.c:
```diff
@@ -2359,16 +2372,35 @@ dummy_func(
+            DEOPT_IF(!LOCK_OBJECT(dict));
+            #ifdef Py_GIL_DISABLED
+            dict = _PyObject_GetManagedDict(owner_o);
+            if (dict == NULL) {
```
This may overwrite the local variable `dict` with a *different* dictionary, which will cause problems for subsequent calls like `UNLOCK_OBJECT(dict)`.

I'm not entirely sure if we need a check list this here, but if we do, I think it should be something like:

```
#ifdef Py_GIL_DISABLED
if (dict != _PyObject_GetManagedDict(owner_o)) {
    UNLOCK_OBJECT(dict);
    DEOPT_IF(true);
}
#endif
```

colesbury commented on Python/bytecodes.c:
```diff
@@ -2128,7 +2128,18 @@ dummy_func(
             PyTypeObject *tp = Py_TYPE(PyStackRef_AsPyObjectBorrow(owner));
             assert(type_version != 0);
-            EXIT_IF(tp->tp_version_tag != type_version);
+            EXIT_IF(FT_ATOMIC_LOAD_UINT32_RELAXED(tp->tp_version_tag) != type_version);
```
The accesses to `tp_version_tag` should use `FT_ATOMIC_LOAD_UINT_RELAXED` instead of `FT_ATOMIC_LOAD_UINT32_RELAXED` because it's an `unsigned int`.

colesbury replied:
Oh... we have a lot of these that should be changed. We can do in a separate PR.

nascheme committed:
Use correct type of load for tp_version_tag.

nascheme committed:
Avoid overwritting 'dict', needed for unlock.

This also fixes the case if the dict is replaced with a different one.

nascheme committed:
Add additional tests for STORE_ATTR.

mpage commented on Python/specialize.c:
```diff
@@ -881,71 +881,142 @@ classify_descriptor(PyObject *descriptor, bool has_getattr)
+    _PyAttrCache *cache = (_PyAttrCache *)(instr + 1);
+    PyDictKeysObject *keys = ((PyHeapTypeObject *)type)->ht_cached_keys;
+    assert(PyUnicode_CheckExact(name));
+    Py_ssize_t index = _PyDictKeys_StringLookup(keys, name);
```
I think we need to lock the keys object in the free-threaded builds using `_Py_LOCK_DONT_DETACH` (see `{LOCK,UNLOCK}_KEYS()` in `Objects/dictobject.c`).

mpage commented on Python/specialize.c:
```diff
@@ -1712,19 +1769,19 @@ function_get_version(PyObject *o, int opcode)
 type_get_version(PyTypeObject *t, int opcode)
 {
-    uint32_t version = t->tp_version_tag;
+    uint32_t version = FT_ATOMIC_LOAD_UINT_RELAXED(t->tp_version_tag);
```
I don't think this is needed: `type_get_version` is only used in the default build.
```suggestion
    uint32_t version = t->tp_version_tag;
```

mpage reviewed (commented):
I left a couple of small comments inline, but LGTM overall.

nascheme committed:
Add locking for split dict keys.

For `specialize_dict_access_inline()`, we need to lock the keys object.

* Add `_PyDictKeys_StringLookupSplit` which does required locking and
  use in place of `_PyDictKeys_StringLookup`.

* Change `_PyObject_TryGetInstanceAttribute` to use that function
  in the case of split keys.

* Add `unicodekeys_lookup_split` helper which allows code sharing
  between `_Py_dict_lookup` and `_PyDictKeys_StringLookupSplit`.

nascheme committed:
Remove unneeded atomic load.

The function is only used in with-GIL builds.

mpage commented on Objects/dictobject.c:
```diff
@@ -1129,6 +1129,38 @@ dictkeys_generic_lookup(PyDictObject *mp, PyDictKeysObject* dk, PyObject *key, P
+    }
+    else {
+        ix = unicodekeys_lookup_unicode(dk, key, hash);
+    }
```
I think we only need to perform the lookup again if the keys changed:
```suggestion
    ix = unicodekeys_lookup_unicode_threadsafe(dk, key, hash);
    if (ix == DKIX_KEY_CHANGED) {
        LOCK_KEYS(dk);
        ix = unicodekeys_lookup_unicode(dk, key, hash);
        UNLOCK_KEYS(dk);
    }
```

mpage commented on Objects/dictobject.c:
```diff
@@ -6967,7 +7002,12 @@ _PyObject_TryGetInstanceAttribute(PyObject *obj, PyObject *name, PyObject **attr
+        ix = _PyDictKeys_StringLookupSplit(keys, name);
+    } else {
+        ix = _PyDictKeys_StringLookup(keys, name);
+    }
```
Since `keys` is always the cached split keys object that is retrieved from the type object, I think this can just be:

```suggestion
    Py_ssize_t ix = _PyDictKeys_StringLookupSplit(keys, name);
```

mpage reviewed (commented):
Nice! Just a couple of small comments inline.

nascheme committed:
Code cleanup, remove unneeded branches.

mpage reviewed (approved):
LGTM! 

Benchmark results:

* [~2% faster](https://github.com/facebookexperimental/free-threading-benchmarking/tree/main/results/bm-20241218-3.14.0a3+-9015a3f-NOGIL#vs-base) on free-threaded builds.
* [~neutral](https://github.com/facebookexperimental/free-threading-benchmarking/tree/main/results/bm-20241218-3.14.0a3+-9015a3f#vs-base) on default builds.

nascheme committed:
Merge branch 'main' into gh-115999-specialize-store-attr

nascheme committed:
Merge branch 'main' into gh-115999-specialize-store-attr
